Can't run Kotlin files in Windows with more than one dependency in the classpath

**Describe the bug**
Can't run Kotlin files in Windows with more than one dependency in the classpath

**To Reproduce**
Steps to reproduce the behavior:
1. Create a `KotlinBug.kt` file with the following content:
```kotlin
///usr/bin/env jbang "$0" "$@" ; exit $?

//DEPS com.squareup.okhttp3:okhttp:4.10.0
//DEPS org.json:json:20220320

import okhttp3.OkHttpClient
import org.json.JSONObject

fun main() {
    val jokeRequest = okhttp3.Request.Builder()
        .url("https://api.chucknorris.io/jokes/random?category=dev")
        .build()
    val jokeResponse = OkHttpClient().newCall(jokeRequest).execute()
    val joke = JSONObject(jokeResponse.body!!.string())
    println(joke["value"])
}
``` 

2. Run it in a `Windows 10` host with `Powershell`, like this:
```ps1
jbang .\KotlinBug.kt
```
3. See the following error:
```ps1
[jbang] Building jar...
error: source entry is not a Kotlin file: C:\Users\gustavo.rodriguezl\.m2\repository\com\squareup\okio\okio-jvm\3.0.0\okio-jvm-3.0.0.jar
error: source entry is not a Kotlin file: C:\Users\gustavo.rodriguezl\.m2\repository\org\jetbrains\kotlin\kotlin-stdlib-jdk8\1.5.31\kotlin-stdlib-jdk8-1.5.31.jar
error: source entry is not a Kotlin file: C:\Users\gustavo.rodriguezl\.m2\repository\org\jetbrains\kotlin\kotlin-stdlib-jdk7\1.5.31\kotlin-stdlib-jdk7-1.5.31.jar
error: source entry is not a Kotlin file: C:\Users\gustavo.rodriguezl\.m2\repository\org\jetbrains\kotlin\kotlin-stdlib-common\1.5.31\kotlin-stdlib-common-1.5.31.jar
error: source entry is not a Kotlin file: C:\Users\gustavo.rodriguezl\.m2\repository\org\jetbrains\kotlin\kotlin-stdlib\1.6.20\kotlin-stdlib-1.6.20.jar
error: source entry is not a Kotlin file: C:\Users\gustavo.rodriguezl\.m2\repository\org\jetbrains\annotations\13.0\annotations-13.0.jar
error: source entry is not a Kotlin file: C:\Users\gustavo.rodriguezl\.m2\repository\org\json\json\20220320\json-20220320.jar
[jbang] [ERROR] Error during compile
[jbang] Run with --verbose for more details
```

**Expected behavior**
- The script should just run fine as it does in an `Unix` host:
```bash
❯ jbang KotlinBug.kt
[jbang] Building jar...
When Chuck Norris break the build, you can't fix it, because there is not a single line of code left.
```

**JBang version**
```ps1
[jbang] jbang version 0.98.0
Cache: C:\Users\gustavo.rodriguezl\.jbang\cache
Config: C:\Users\gustavo.rodriguezl\.jbang
Repository:C:\Users\gustavo.rodriguezl\.m2\repository
0.98.0
```
**Additional context**

👀👇👇

<details>
  <summary>This is how the execution looks if verbose mode is enabled</summary>

  ```ps1
    gustavo.rodriguezl ~\src\garodriguezlp\jbang-kotlin-bug
    ❯ jbang --verbose run .\KotlinBug.kt
    [jbang] jbang version 0.98.0
    [jbang] Build required as C:\Users\gustavo.rodriguezl\.jbang\cache\jars\KotlinBug.kt.b1dcd226c361b16405f16afa01abf35b4bc837922d85cd9439255613c8efff4d.jar not readable or not found.
    [jbang] Downloading Kotlin 1.6.10. Be patient, this can take several minutes...
    [jbang] Downloading https://github.com/JetBrains/kotlin/releases/download/v1.6.10/kotlin-compiler-1.6.10.zip
    [jbang] Redirected to: https://objects.githubusercontent.com/github-production-release-asset-2e65be/3432266/321575a3-fab9-4e5a-9a47-dda6e8303e47?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20220902%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20220902T191331Z&X-Amz-Expires=300&X-Amz-Signature=429d8a730640798c6f4318335b48b4313757750623969938dc0e41eec48ac37c&X-Amz-SignedHeaders=host&actor_id=0&key_id=0&repo_id=3432266&response-content-disposition=attachment%3B%20filename%3Dkotlin-compiler-1.6.10.zip&response-content-type=application%2Foctet-stream
    [jbang] Downloaded file https://objects.githubusercontent.com/github-production-release-asset-2e65be/3432266/321575a3-fab9-4e5a-9a47-dda6e8303e47?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20220902%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20220902T191331Z&X-Amz-Expires=300&X-Amz-Signature=429d8a730640798c6f4318335b48b4313757750623969938dc0e41eec48ac37c&X-Amz-SignedHeaders=host&actor_id=0&key_id=0&repo_id=3432266&response-content-disposition=attachment%3B%20filename%3Dkotlin-compiler-1.6.10.zip&response-content-type=application%2Foctet-stream
    [jbang] Installing Kotlin 1.6.10...
    [jbang] Unpacking to C:\Users\gustavo.rodriguezl\.jbang\cache\kotlincs\1.6.10
    [jbang] Repositories: mavencentral=https://repo1.maven.org/maven2/
    [jbang] Resolving dependencies...
    [jbang] com.squareup.okhttp3:okhttp:jar:4.10.0
            org.json:json:jar:20220320
    Done
    [jbang] Dependencies resolved
    [jbang] Building jar...
    [jbang] Compile: C:\Users\gustavo.rodriguezl\.jbang\cache\kotlincs\1.6.10\kotlinc\bin\kotlinc.bat -classpath C:\Users\gustavo.rodriguezl\.m2\repository\com\squareup\okhttp3\okhttp\4.10.0\okhttp-4.10.0.jar;C:\Users\gustavo.rodriguezl\.m2\repository\com\squareup\okio\okio-jvm\3.0.0\okio-jvm-3.0.0.jar;C:\Users\gustavo.rodriguezl\.m2\repository\org\jetbrains\kotlin\kotlin-stdlib-jdk8\1.5.31\kotlin-stdlib-jdk8-1.5.31.jar;C:\Users\gustavo.rodriguezl\.m2\repository\org\jetbrains\kotlin\kotlin-stdlib-jdk7\1.5.31\kotlin-stdlib-jdk7-1.5.31.jar;C:\Users\gustavo.rodriguezl\.m2\repository\org\jetbrains\kotlin\kotlin-stdlib-common\1.5.31\kotlin-stdlib-common-1.5.31.jar;C:\Users\gustavo.rodriguezl\.m2\repository\org\jetbrains\kotlin\kotlin-stdlib\1.6.20\kotlin-stdlib-1.6.20.jar;C:\Users\gustavo.rodriguezl\.m2\repository\org\jetbrains\annotations\13.0\annotations-13.0.jar;C:\Users\gustavo.rodriguezl\.m2\repository\org\json\json\20220320\json-20220320.jar -d C:\Users\gustavo.rodriguezl\.jbang\cache\jars\KotlinBug.kt.b1dcd226c361b16405f16afa01abf35b4bc837922d85cd9439255613c8efff4d.jar.tmp C:\Users\gustavo.rodriguezl\src\garodriguezlp\jbang-kotlin-bug\KotlinBug.kt
    error: source entry is not a Kotlin file: C:\Users\gustavo.rodriguezl\.m2\repository\com\squareup\okio\okio-jvm\3.0.0\okio-jvm-3.0.0.jar
    error: source entry is not a Kotlin file: C:\Users\gustavo.rodriguezl\.m2\repository\org\jetbrains\kotlin\kotlin-stdlib-jdk8\1.5.31\kotlin-stdlib-jdk8-1.5.31.jar
    error: source entry is not a Kotlin file: C:\Users\gustavo.rodriguezl\.m2\repository\org\jetbrains\kotlin\kotlin-stdlib-jdk7\1.5.31\kotlin-stdlib-jdk7-1.5.31.jar
    error: source entry is not a Kotlin file: C:\Users\gustavo.rodriguezl\.m2\repository\org\jetbrains\kotlin\kotlin-stdlib-common\1.5.31\kotlin-stdlib-common-1.5.31.jar
    error: source entry is not a Kotlin file: C:\Users\gustavo.rodriguezl\.m2\repository\org\jetbrains\kotlin\kotlin-stdlib\1.6.20\kotlin-stdlib-1.6.20.jar
    error: source entry is not a Kotlin file: C:\Users\gustavo.rodriguezl\.m2\repository\org\jetbrains\annotations\13.0\annotations-13.0.jar
    error: source entry is not a Kotlin file: C:\Users\gustavo.rodriguezl\.m2\repository\org\json\json\20220320\json-20220320.jar
    [jbang] Deleting folder C:\Users\gustavo.rodriguezl\.jbang\cache\jars\KotlinBug.kt.b1dcd226c361b16405f16afa01abf35b4bc837922d85cd9439255613c8efff4d.jar.tmp
    [jbang] [ERROR] Error during compile
    dev.jbang.cli.ExitException: Error during compile
            at dev.jbang.source.builders.BaseBuilder.runCompiler(BaseBuilder.java:191)
            at dev.jbang.source.builders.BaseBuilder.runCompiler(BaseBuilder.java:179)
            at dev.jbang.source.builders.BaseBuilder.compile(BaseBuilder.java:152)
            at dev.jbang.source.builders.BaseBuilder.build(BaseBuilder.java:105)
            at dev.jbang.source.builders.BaseBuilder.build(BaseBuilder.java:30)
            at dev.jbang.cli.Run.prepareArtifacts(Run.java:133)
            at dev.jbang.cli.Run.doCall(Run.java:104)
            at dev.jbang.cli.BaseCommand.call(BaseCommand.java:136)
            at dev.jbang.cli.BaseCommand.call(BaseCommand.java:20)
            at picocli.CommandLine.executeUserObject(CommandLine.java:1953)
            at picocli.CommandLine.access$1300(CommandLine.java:145)
            at picocli.CommandLine$RunLast.executeUserObjectOfLastSubcommandWithSameParent(CommandLine.java:2358)
            at picocli.CommandLine$RunLast.handle(CommandLine.java:2352)
            at dev.jbang.cli.JBang$3.handle(JBang.java:151)
            at dev.jbang.cli.JBang$3.handle(JBang.java:146)
            at picocli.CommandLine$AbstractParseResultHandler.execute(CommandLine.java:2179)
            at picocli.CommandLine$RunLast.execute(CommandLine.java:2316)
            at picocli.CommandLine.execute(CommandLine.java:2078)
            at dev.jbang.Main.main(Main.java:14)
    [jbang] If you believe this a bug in jbang, open an issue at https://github.com/jbangdev/jbang/issues
  ```
</details>

<details>
  <summary>And this is what I see if I dig into and tweek the `kotlinc.bat` file</summary>


  - Actually, I'm talking about modifying the [`kotlinc.bat` file line 69](https://github.com/JetBrains/kotlin/blob/master/compiler/cli/bin/kotlinc.bat#L69) by puttig before and `echo` command to see what is the value of the `KOTLIN_OPTS` variable.
    - Like this: 
    ```bat
    echo "KOTLIN_OPTS=%KOTLIN_OPTS%"
    ```

  ```ps1
    ❯ jbang --verbose run .\KotlinBug.kt
    [jbang] jbang version 0.98.0
    [jbang] Build required as C:\Users\gustavo.rodriguezl\.jbang\cache\jars\KotlinBug.kt.b1dcd226c361b16405f16afa01abf35b4bc837922d85cd9439255613c8efff4d.jar not readable or not found.
    [jbang] Repositories: mavencentral=https://repo1.maven.org/maven2/
    [jbang] Building jar...
    [jbang] Compile: C:\Users\gustavo.rodriguezl\.jbang\cache\kotlincs\1.6.10\kotlinc\bin\kotlinc.bat -classpath C:\Users\gustavo.rodriguezl\.m2\repository\com\squareup\okhttp3\okhttp\4.10.0\okhttp-4.10.0.jar;C:\Users\gustavo.rodriguezl\.m2\repository\com\squareup\okio\okio-jvm\3.0.0\okio-jvm-3.0.0.jar;C:\Users\gustavo.rodriguezl\.m2\repository\org\jetbrains\kotlin\kotlin-stdlib-jdk8\1.5.31\kotlin-stdlib-jdk8-1.5.31.jar;C:\Users\gustavo.rodriguezl\.m2\repository\org\jetbrains\kotlin\kotlin-stdlib-jdk7\1.5.31\kotlin-stdlib-jdk7-1.5.31.jar;C:\Users\gustavo.rodriguezl\.m2\repository\org\jetbrains\kotlin\kotlin-stdlib-common\1.5.31\kotlin-stdlib-common-1.5.31.jar;C:\Users\gustavo.rodriguezl\.m2\repository\org\jetbrains\kotlin\kotlin-stdlib\1.6.20\kotlin-stdlib-1.6.20.jar;C:\Users\gustavo.rodriguezl\.m2\repository\org\jetbrains\annotations\13.0\annotations-13.0.jar;C:\Users\gustavo.rodriguezl\.m2\repository\org\json\json\20220320\json-20220320.jar -d C:\Users\gustavo.rodriguezl\.jbang\cache\jars\KotlinBug.kt.b1dcd226c361b16405f16afa01abf35b4bc837922d85cd9439255613c8efff4d.jar.tmp C:\Users\gustavo.rodriguezl\src\garodriguezlp\jbang-kotlin-bug\KotlinBug.kt
    error: source entry is not a Kotlin file: C:\Users\gustavo.rodriguezl\.m2\repository\com\squareup\okio\okio-jvm\3.0.0\okio-jvm-3.0.0.jar
    error: source entry is not a Kotlin file: C:\Users\gustavo.rodriguezl\.m2\repository\org\jetbrains\kotlin\kotlin-stdlib-jdk8\1.5.31\kotlin-stdlib-jdk8-1.5.31.jar
    error: source entry is not a Kotlin file: C:\Users\gustavo.rodriguezl\.m2\repository\org\jetbrains\kotlin\kotlin-stdlib-jdk7\1.5.31\kotlin-stdlib-jdk7-1.5.31.jar
    error: source entry is not a Kotlin file: C:\Users\gustavo.rodriguezl\.m2\repository\org\jetbrains\kotlin\kotlin-stdlib-common\1.5.31\kotlin-stdlib-common-1.5.31.jar
    error: source entry is not a Kotlin file: C:\Users\gustavo.rodriguezl\.m2\repository\org\jetbrains\kotlin\kotlin-stdlib\1.6.20\kotlin-stdlib-1.6.20.jar
    error: source entry is not a Kotlin file: C:\Users\gustavo.rodriguezl\.m2\repository\org\jetbrains\annotations\13.0\annotations-13.0.jar
    error: source entry is not a Kotlin file: C:\Users\gustavo.rodriguezl\.m2\repository\org\json\json\20220320\json-20220320.jar
    [jbang] Deleting folder C:\Users\gustavo.rodriguezl\.jbang\cache\jars\KotlinBug.kt.b1dcd226c361b16405f16afa01abf35b4bc837922d85cd9439255613c8efff4d.jar.tmp
    [jbang] [ERROR] Error during compile
    dev.jbang.cli.ExitException: Error during compile
            at dev.jbang.source.builders.BaseBuilder.runCompiler(BaseBuilder.java:191)
            at dev.jbang.source.builders.BaseBuilder.runCompiler(BaseBuilder.java:179)
            at dev.jbang.source.builders.BaseBuilder.compile(BaseBuilder.java:152)
            at dev.jbang.source.builders.BaseBuilder.build(BaseBuilder.java:105)
            at dev.jbang.source.builders.BaseBuilder.build(BaseBuilder.java:30)
            at dev.jbang.cli.Run.prepareArtifacts(Run.java:133)
            at dev.jbang.cli.Run.doCall(Run.java:104)
            at dev.jbang.cli.BaseCommand.call(BaseCommand.java:136)
            at dev.jbang.cli.BaseCommand.call(BaseCommand.java:20)
            at picocli.CommandLine.executeUserObject(CommandLine.java:1953)
            at picocli.CommandLine.access$1300(CommandLine.java:145)
            at picocli.CommandLine$RunLast.executeUserObjectOfLastSubcommandWithSameParent(CommandLine.java:2358)
            at picocli.CommandLine$RunLast.handle(CommandLine.java:2352)
            at dev.jbang.cli.JBang$3.handle(JBang.java:151)
            at dev.jbang.cli.JBang$3.handle(JBang.java:146)
            at picocli.CommandLine$AbstractParseResultHandler.execute(CommandLine.java:2179)
            at picocli.CommandLine$RunLast.execute(CommandLine.java:2316)
            at picocli.CommandLine.execute(CommandLine.java:2078)
            at dev.jbang.Main.main(Main.java:14)
    [jbang] If you believe this a bug in jbang, open an issue at https://github.com/jbangdev/jbang/issues
    "KOTLIN_OPTS= "-classpath" "C:\Users\gustavo.rodriguezl\.m2\repository\com\squareup\okhttp3\okhttp\4.10.0\okhttp-4.10.0.jar" "C:\Users\gustavo.rodriguezl\.m2\repository\com\squareup\okio\okio-jvm\3.0.0\okio-jvm-3.0.0.jar" "C:\Users\gustavo.rodriguezl\.m2\repository\org\jetbrains\kotlin\kotlin-stdlib-jdk8\1.5.31\kotlin-stdlib-jdk8-1.5.31.jar" "C:\Users\gustavo.rodriguezl\.m2\repository\org\jetbrains\kotlin\kotlin-stdlib-jdk7\1.5.31\kotlin-stdlib-jdk7-1.5.31.jar" "C:\Users\gustavo.rodriguezl\.m2\repository\org\jetbrains\kotlin\kotlin-stdlib-common\1.5.31\kotlin-stdlib-common-1.5.31.jar" "C:\Users\gustavo.rodriguezl\.m2\repository\org\jetbrains\kotlin\kotlin-stdlib\1.6.20\kotlin-stdlib-1.6.20.jar" "C:\Users\gustavo.rodriguezl\.m2\repository\org\jetbrains\annotations\13.0\annotations-13.0.jar" "C:\Users\gustavo.rodriguezl\.m2\repository\org\json\json\20220320\json-20220320.jar" "-d" "C:\Users\gustavo.rodriguezl\.jbang\cache\jars\KotlinBug.kt.b1dcd226c361b16405f16afa01abf35b4bc837922d85cd9439255613c8efff4d.jar.tmp" "C:\Users\gustavo.rodriguezl\src\garodriguezlp\jbang-kotlin-bug\KotlinBug.kt""
  ```

  - As you can see (in this👆last debug line), the classpath is not being passed correctly to the Kotlin compiler. The classpath is being passed as multiple arguments instead of a single argument separated by `;` semicolon.

</details>

--- a/itests/echo.bat
+++ b/itests/echo.bat
@@ -1,0 +1,10 @@
+@echo off
+setlocal
+echo ARGS = %*
+:loop
+set _arg=%~1
+if "%_arg%" == "" goto loopend
+echo ARG = %_arg%
+shift
+goto loop
+:loopend

--- a/src/main/java/dev/jbang/cli/Edit.java
+++ b/src/main/java/dev/jbang/cli/Edit.java
@@ -1,7 +1,6 @@
 package dev.jbang.cli;
 
 import static dev.jbang.Settings.CP_SEPARATOR;
-import static dev.jbang.source.builders.BaseBuilder.escapeOSArguments;
 import static dev.jbang.util.Util.verboseMsg;
 import static java.lang.System.out;
 
@@ -26,6 +25,7 @@ import dev.jbang.dependencies.JitPackUtil;
 import dev.jbang.dependencies.MavenRepo;
 import dev.jbang.net.EditorManager;
 import dev.jbang.source.*;
+import dev.jbang.util.CommandBuffer;
 import dev.jbang.util.TemplateEngine;
 import dev.jbang.util.Util;
 import dev.jbang.util.Util.Shell;
@@ -155,10 +155,10 @@ public class Edit extends BaseCommand {
 
 			String[] cmd;
 			if (Util.getShell() == Shell.bash) {
-				final String editorCommand = String.join(" ", escapeOSArguments(optionList, Shell.bash));
+				final String editorCommand = CommandBuffer.of(optionList).asCommandLine(Shell.bash);
 				cmd = new String[] { "sh", "-c", editorCommand };
 			} else {
-				final String editorCommand = String.join(" ", escapeOSArguments(optionList, Shell.cmd));
+				final String editorCommand = CommandBuffer.of(optionList).asCommandLine(Shell.cmd);
 				cmd = new String[] { "cmd", "/c", editorCommand };
 			}
 			verboseMsg("Running `" + String.join(" ", cmd) + "`");

--- a/src/main/java/dev/jbang/source/builders/BaseBuilder.java
+++ b/src/main/java/dev/jbang/source/builders/BaseBuilder.java
@@ -8,7 +8,6 @@ import java.util.*;
 import java.util.function.Predicate;
 import java.util.jar.Attributes;
 import java.util.jar.Manifest;
-import java.util.regex.Pattern;
 import java.util.stream.Collectors;
 import java.util.stream.Stream;
 
@@ -20,10 +19,7 @@ import dev.jbang.dependencies.DependencyUtil;
 import dev.jbang.source.*;
 import dev.jbang.spi.IntegrationManager;
 import dev.jbang.spi.IntegrationResult;
-import dev.jbang.util.JarUtil;
-import dev.jbang.util.JavaUtil;
-import dev.jbang.util.TemplateEngine;
-import dev.jbang.util.Util;
+import dev.jbang.util.*;
 
 import io.quarkus.qute.Template;
 
@@ -176,7 +172,7 @@ public abstract class BaseBuilder implements Builder {
 	}
 
 	protected void runCompiler(List<String> optionList) throws IOException {
-		runCompiler(new ProcessBuilder(optionList).inheritIO());
+		runCompiler(CommandBuffer.of(optionList).asProcessBuilder().inheritIO());
 	}
 
 	protected void runCompiler(ProcessBuilder processBuilder) throws IOException {
@@ -210,7 +206,7 @@ public abstract class BaseBuilder implements Builder {
 		// options set on the Source (that way persistent args can override
 		// options set on the Source)
 		List<String> rtArgs = prj.getRuntimeOptions();
-		String runtimeOpts = String.join(" ", escapeArguments(rtArgs));
+		String runtimeOpts = CommandBuffer.of(rtArgs).asCommandLine(Util.Shell.bash);
 		if (!runtimeOpts.isEmpty()) {
 			manifest.getMainAttributes()
 					.putValue(ATTR_JBANG_JAVA_OPTIONS, runtimeOpts);
@@ -270,7 +266,11 @@ public abstract class BaseBuilder implements Builder {
 		Util.verboseMsg("native-image: " + String.join(" ", optionList));
 		Util.infoMsg("log: " + nilog.toString());
 
-		Process process = new ProcessBuilder(optionList).inheritIO().redirectOutput(nilog.toFile()).start();
+		Process process = CommandBuffer	.of(optionList)
+										.asProcessBuilder()
+										.inheritIO()
+										.redirectOutput(nilog.toFile())
+										.start();
 		try {
 			process.waitFor();
 		} catch (InterruptedException e) {
@@ -314,74 +314,6 @@ public abstract class BaseBuilder implements Builder {
 		} else {
 			return cmd;
 		}
-	}
-
-	// NB: This might not be a definitive list of safe characters
-	static Pattern cmdSafeChars = Pattern.compile("[a-zA-Z0-9.,_+=:;@()-]*");
-	// TODO: Figure out what the real list of safe characters is for PowerShell
-	static Pattern pwrSafeChars = Pattern.compile("[a-zA-Z0-9.,_+=:;@()-]*");
-	static Pattern shellSafeChars = Pattern.compile("[a-zA-Z0-9._+=:@%/-]*");
-
-	/**
-	 * Escapes list of arguments where necessary using the current OS' way of
-	 * escaping
-	 */
-	public static List<String> escapeOSArguments(List<String> args, Util.Shell shell) {
-		return args.stream().map(arg -> escapeOSArgument(arg, shell)).collect(Collectors.toList());
-	}
-
-	/**
-	 * Escapes list of arguments where necessary using a generic way of escaping
-	 * (we'll just be using the Unix way)
-	 */
-	static List<String> escapeArguments(List<String> args) {
-		return args.stream().map(BaseBuilder::escapeUnixArgument).collect(Collectors.toList());
-	}
-
-	public static String escapeOSArgument(String arg, Util.Shell shell) {
-		switch (shell) {
-		case bash:
-			return escapeUnixArgument(arg);
-		case cmd:
-			return escapeCmdArgument(arg);
-		case powershell:
-			return escapePowershellArgument(arg);
-		}
-		return arg;
-	}
-
-	static String escapeUnixArgument(String arg) {
-		if (!shellSafeChars.matcher(arg).matches()) {
-			arg = arg.replaceAll("(['])", "'\\\\''");
-			arg = "'" + arg + "'";
-		}
-		return arg;
-	}
-
-	public static String escapeArgsFileArgument(String arg) {
-		if (!shellSafeChars.matcher(arg).matches()) {
-			arg = arg.replaceAll("([\"'\\\\])", "\\\\$1");
-			arg = "\"" + arg + "\"";
-		}
-		return arg;
-	}
-
-	static String escapeCmdArgument(String arg) {
-		if (!cmdSafeChars.matcher(arg).matches()) {
-			// Windows quoting is just weird
-			arg = arg.replaceAll("([()!^<>&|% ])", "^$1");
-			arg = arg.replaceAll("([\"])", "\\\\^$1");
-			arg = "^\"" + arg + "^\"";
-		}
-		return arg;
-	}
-
-	static String escapePowershellArgument(String arg) {
-		if (!pwrSafeChars.matcher(arg).matches()) {
-			arg = arg.replaceAll("(['])", "''");
-			arg = "'" + arg + "'";
-		}
-		return arg;
 	}
 
 	protected void searchForMain(Path tmpJarDir) {

--- a/src/main/java/dev/jbang/source/generators/BaseCmdGenerator.java
+++ b/src/main/java/dev/jbang/source/generators/BaseCmdGenerator.java
@@ -3,14 +3,12 @@ package dev.jbang.source.generators;
 import static dev.jbang.source.builders.BaseBuilder.*;
 
 import java.io.IOException;
-import java.io.PrintWriter;
-import java.nio.file.Files;
-import java.nio.file.Path;
 import java.util.*;
 
 import dev.jbang.cli.BaseCommand;
 import dev.jbang.cli.ExitException;
 import dev.jbang.source.*;
+import dev.jbang.util.CommandBuffer;
 import dev.jbang.util.JavaUtil;
 import dev.jbang.util.Util;
 
@@ -34,7 +32,8 @@ public abstract class BaseCmdGenerator implements CmdGenerator {
 	@Override
 	public String generate() throws IOException {
 		List<String> fullArgs = generateCommandLineList();
-		String args = String.join(" ", escapeOSArguments(fullArgs, shell));
+		CommandBuffer cb = CommandBuffer.of(fullArgs);
+		String args = cb.asCommandLine(shell);
 		// Check if we can and need to use @-files on Windows
 		boolean useArgsFile = false;
 		if (!(getCode().isJShell() || ctx.getForceType() == Source.Type.jshell) &&
@@ -45,17 +44,7 @@ public abstract class BaseCmdGenerator implements CmdGenerator {
 			useArgsFile = actualVersion >= 9;
 		}
 		if (useArgsFile) {
-			// @-files avoid problems on Windows with very long command lines
-			final String javaCmd = escapeOSArgument(fullArgs.get(0), shell);
-			final Path argsFile = Files.createTempFile("jbang", ".args");
-			try (PrintWriter pw = new PrintWriter(argsFile.toFile())) {
-				// write all arguments except the first to the file
-				for (int i = 1; i < fullArgs.size(); ++i) {
-					pw.println(escapeArgsFileArgument(fullArgs.get(i)));
-				}
-			}
-
-			return javaCmd + " @" + argsFile;
+			return cb.asJavaArgsFile(shell);
 		} else {
 			return args;
 		}

--- a/src/main/java/dev/jbang/util/CommandBuffer.java
+++ b/src/main/java/dev/jbang/util/CommandBuffer.java
@@ -15,12 +15,13 @@ public class CommandBuffer {
 	private List<String> arguments;
 
 	// NB: This might not be a definitive list of safe characters
-	// static Pattern cmdSafeChars = Pattern.compile("[^&()\\[]\\{}\\^=;!'+,`~<>|
-	// ]*");
+	// cmdSafeChars = Pattern.compile("[^\\Q&()[]{}^=;!'+,`~<>|\\E]*");
 	static Pattern cmdSafeChars = Pattern.compile("[a-zA-Z0-9.,_+=:;@()-]*");
 	// TODO: Figure out what the real list of safe characters is for PowerShell
 	static Pattern pwrSafeChars = Pattern.compile("[a-zA-Z0-9.,_+=:;@()-]*");
 	static Pattern shellSafeChars = Pattern.compile("[a-zA-Z0-9._+=:@%/-]*");
+
+	static Pattern cmdNeedQuotesChars = Pattern.compile("[\\Q&()[]{}^=;!'+,`~\\E]");
 
 	public static CommandBuffer of() {
 		return new CommandBuffer();
@@ -47,7 +48,14 @@ public class CommandBuffer {
 	}
 
 	public ProcessBuilder asProcessBuilder() {
-		return new ProcessBuilder(arguments);
+		return asProcessBuilder(Util.getShell());
+	}
+
+	public ProcessBuilder asProcessBuilder(Util.Shell shell) {
+		List<String> args = arguments	.stream()
+										.map(a -> escapeProcessBuilderArgument(a, shell))
+										.collect(Collectors.toList());
+		return new ProcessBuilder(args);
 	}
 
 	public String asCommandLine() {
@@ -126,8 +134,15 @@ public class CommandBuffer {
 	}
 
 	private static String escapeProcessBuilderArgument(String arg, Util.Shell shell) {
-		if (shell != Util.Shell.bash) {
-			arg = escapeCmdArgument(arg);
+		// This code is meant to deal with special characters on Windows
+		// that get treated as argument separators if left unquoted.
+		// (For a list of those characters see the output of `CMD /?`)
+		// The code is NOT able to handle double quotes as part of the
+		// argument, this is close to impossible to handle on Windows.
+		// (see
+		// https://stackoverflow.com/questions/30157414/batch-argument-with-quotes-and-spaces)
+		if (shell != Util.Shell.bash && cmdNeedQuotesChars.matcher(arg).find()) {
+			arg = "\"" + arg + "\"";
 		}
 		return arg;
 	}

--- a/src/main/java/dev/jbang/util/CommandBuffer.java
+++ b/src/main/java/dev/jbang/util/CommandBuffer.java
@@ -1,0 +1,134 @@
+package dev.jbang.util;
+
+import java.io.IOException;
+import java.io.PrintWriter;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.List;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+public class CommandBuffer {
+	private List<String> arguments;
+
+	// NB: This might not be a definitive list of safe characters
+	// static Pattern cmdSafeChars = Pattern.compile("[^&()\\[]\\{}\\^=;!'+,`~<>|
+	// ]*");
+	static Pattern cmdSafeChars = Pattern.compile("[a-zA-Z0-9.,_+=:;@()-]*");
+	// TODO: Figure out what the real list of safe characters is for PowerShell
+	static Pattern pwrSafeChars = Pattern.compile("[a-zA-Z0-9.,_+=:;@()-]*");
+	static Pattern shellSafeChars = Pattern.compile("[a-zA-Z0-9._+=:@%/-]*");
+
+	public static CommandBuffer of() {
+		return new CommandBuffer();
+	}
+
+	public static CommandBuffer of(Collection<String> arguments) {
+		return new CommandBuffer(arguments);
+	}
+
+	public static CommandBuffer of(String... arguments) {
+		return new CommandBuffer(arguments);
+	}
+
+	public CommandBuffer() {
+		arguments = new ArrayList<>();
+	}
+
+	public CommandBuffer(Collection<String> arguments) {
+		this.arguments = new ArrayList<>(arguments);
+	}
+
+	public CommandBuffer(String... arguments) {
+		this.arguments = new ArrayList<>(Arrays.asList(arguments));
+	}
+
+	public ProcessBuilder asProcessBuilder() {
+		return new ProcessBuilder(arguments);
+	}
+
+	public String asCommandLine() {
+		return asCommandLine(Util.getShell());
+	}
+
+	public String asCommandLine(Util.Shell shell) {
+		return String.join(" ", escapeShellArguments(arguments, shell));
+	}
+
+	public String asJavaArgsFile(Util.Shell shell) throws IOException {
+		// @-files avoid problems on Windows with very long command lines
+		final String cmd = escapeShellArgument(arguments.get(0), shell);
+		final Path argsFile = Files.createTempFile("jbang", ".args");
+		try (PrintWriter pw = new PrintWriter(argsFile.toFile())) {
+			// write all arguments except the first to the file
+			for (int i = 1; i < arguments.size(); ++i) {
+				pw.println(escapeArgsFileArgument(arguments.get(i)));
+			}
+		}
+		return cmd + " @" + argsFile;
+	}
+
+	/**
+	 * Escapes list of arguments where necessary using the way of escaping that is
+	 * appropriate for the shell we're running in
+	 */
+	private static List<String> escapeShellArguments(List<String> args, Util.Shell shell) {
+		return args.stream().map(arg -> escapeShellArgument(arg, shell)).collect(Collectors.toList());
+	}
+
+	public static String escapeShellArgument(String arg, Util.Shell shell) {
+		switch (shell) {
+		case bash:
+			return escapeBashArgument(arg);
+		case cmd:
+			return escapeCmdArgument(arg);
+		case powershell:
+			return escapePowershellArgument(arg);
+		}
+		return arg;
+	}
+
+	private static String escapeBashArgument(String arg) {
+		if (!shellSafeChars.matcher(arg).matches()) {
+			arg = arg.replaceAll("(['])", "'\\\\''");
+			arg = "'" + arg + "'";
+		}
+		return arg;
+	}
+
+	private static String escapeCmdArgument(String arg) {
+		if (!cmdSafeChars.matcher(arg).matches()) {
+			// Windows quoting is just weird
+			arg = arg.replaceAll("([()!^<>&|% ])", "^$1");
+			arg = arg.replaceAll("([\"])", "\\\\^$1");
+			arg = "^\"" + arg + "^\"";
+		}
+		return arg;
+	}
+
+	private static String escapePowershellArgument(String arg) {
+		if (!pwrSafeChars.matcher(arg).matches()) {
+			arg = arg.replaceAll("(['])", "''");
+			arg = "'" + arg + "'";
+		}
+		return arg;
+	}
+
+	private static String escapeArgsFileArgument(String arg) {
+		if (!shellSafeChars.matcher(arg).matches()) {
+			arg = arg.replaceAll("([\"'\\\\])", "\\\\$1");
+			arg = "\"" + arg + "\"";
+		}
+		return arg;
+	}
+
+	private static String escapeProcessBuilderArgument(String arg, Util.Shell shell) {
+		if (shell != Util.Shell.bash) {
+			arg = escapeCmdArgument(arg);
+		}
+		return arg;
+	}
+}

--- a/src/main/java/dev/jbang/util/Util.java
+++ b/src/main/java/dev/jbang/util/Util.java
@@ -1053,7 +1053,7 @@ public class Util {
 			pb.redirectErrorStream(true);
 			Process p = pb.start();
 			BufferedReader br = new BufferedReader(new InputStreamReader(p.getInputStream()));
-			String cmdOutput = br.lines().collect(Collectors.joining());
+			String cmdOutput = br.lines().collect(Collectors.joining("\n"));
 			int exitCode = p.waitFor();
 			if (exitCode == 0) {
 				return cmdOutput;

--- a/src/main/java/dev/jbang/util/Util.java
+++ b/src/main/java/dev/jbang/util/Util.java
@@ -1049,7 +1049,7 @@ public class Util {
 	 */
 	public static String runCommand(String... cmd) {
 		try {
-			ProcessBuilder pb = new ProcessBuilder(cmd);
+			ProcessBuilder pb = CommandBuffer.of(cmd).asProcessBuilder();
 			pb.redirectErrorStream(true);
 			Process p = pb.start();
 			BufferedReader br = new BufferedReader(new InputStreamReader(p.getInputStream()));

--- a/src/test/java/dev/jbang/cli/TestRun.java
+++ b/src/test/java/dev/jbang/cli/TestRun.java
@@ -65,6 +65,7 @@ import dev.jbang.source.generators.JarCmdGenerator;
 import dev.jbang.source.generators.JshCmdGenerator;
 import dev.jbang.source.resolvers.LiteralScriptResourceResolver;
 import dev.jbang.source.sources.JavaSource;
+import dev.jbang.util.CommandBuffer;
 import dev.jbang.util.JavaUtil;
 import dev.jbang.util.Util;
 
@@ -112,7 +113,7 @@ public class TestRun extends BaseTest {
 		assertThat(result, containsString("picocli-4.6.3.jar"));
 		assertThat(result, containsString("hellojar.jar"));
 		assertThat(result, containsString("-Dfoo=bar"));
-		assertThat(result, containsString(BaseBuilder.escapeOSArgument("-Dbar=aap noot mies", Util.getShell())));
+		assertThat(result, containsString(CommandBuffer.escapeShellArgument("-Dbar=aap noot mies", Util.getShell())));
 		// Make sure the opts only appear once
 		assertThat(result.replaceFirst(Pattern.quote("-Dfoo=bar"), ""),
 				not(containsString("-Dfoo=bar")));
@@ -120,7 +121,8 @@ public class TestRun extends BaseTest {
 				not(containsString("-Dbar=aap noot mies")));
 		// Make sure the opts only appear unquoted
 		assertThat(result,
-				not(containsString(BaseBuilder.escapeOSArgument("-Dfoo=bar -Dbar=aap noot mies", Util.getShell()))));
+				not(containsString(
+						CommandBuffer.escapeShellArgument("-Dfoo=bar -Dbar=aap noot mies", Util.getShell()))));
 		// assertThat(result, containsString("--source 11"));
 	}
 
@@ -143,7 +145,7 @@ public class TestRun extends BaseTest {
 		assertThat(result, containsString("classpath"));
 		assertThat(result, containsString(".jar"));
 		assertThat(result, containsString("-Dfoo=bar"));
-		assertThat(result, containsString(BaseBuilder.escapeOSArgument("-Dbar=aap noot mies", Util.getShell())));
+		assertThat(result, containsString(CommandBuffer.escapeShellArgument("-Dbar=aap noot mies", Util.getShell())));
 		assertThat(result, containsString("-showversion"));
 		// Make sure the opts only appear once
 		assertThat(result.replaceFirst(Pattern.quote("-Dfoo=bar"), ""),
@@ -154,7 +156,8 @@ public class TestRun extends BaseTest {
 				not(containsString("-showversion")));
 		// Make sure the opts only appear unquoted
 		assertThat(result,
-				not(containsString(BaseBuilder.escapeOSArgument("-Dfoo=bar -Dbar=aap noot mies", Util.getShell()))));
+				not(containsString(
+						CommandBuffer.escapeShellArgument("-Dfoo=bar -Dbar=aap noot mies", Util.getShell()))));
 		// assertThat(result, containsString("--source 11"));
 	}
 
@@ -2146,7 +2149,7 @@ public class TestRun extends BaseTest {
 		assertThat(result, matchesRegex(".*build\\.jbang\\.[a-z0-9]+\\.jar.*"));
 		assertThat(result, containsString("picocli-4.6.3.jar"));
 		assertThat(result, containsString("-Dfoo=bar"));
-		assertThat(result, containsString(BaseBuilder.escapeOSArgument("-Dbar=aap noot mies", Util.getShell())));
+		assertThat(result, containsString(CommandBuffer.escapeShellArgument("-Dbar=aap noot mies", Util.getShell())));
 		// Make sure the opts only appear once
 		assertThat(result.replaceFirst(Pattern.quote("-Dfoo=bar"), ""),
 				not(containsString("-Dfoo=bar")));
@@ -2154,7 +2157,8 @@ public class TestRun extends BaseTest {
 				not(containsString("-Dbar=aap noot mies")));
 		// Make sure the opts only appear unquoted
 		assertThat(result,
-				not(containsString(BaseBuilder.escapeOSArgument("-Dfoo=bar -Dbar=aap noot mies", Util.getShell()))));
+				not(containsString(
+						CommandBuffer.escapeShellArgument("-Dfoo=bar -Dbar=aap noot mies", Util.getShell()))));
 		// assertThat(result, containsString("--source 11"));
 	}
 }

--- a/src/test/java/dev/jbang/util/TestCommandBuffer.java
+++ b/src/test/java/dev/jbang/util/TestCommandBuffer.java
@@ -1,0 +1,19 @@
+package dev.jbang.util;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.containsString;
+
+import org.junit.jupiter.api.Test;
+
+import dev.jbang.BaseTest;
+
+public class TestCommandBuffer extends BaseTest {
+	@Test
+	void testRunWinBat() {
+		if (Util.getOS() == Util.OS.windows) {
+			String out = Util.runCommand(examplesTestFolder.resolve("echo.bat").toString(), "abc def", "abc;def");
+			assertThat(out, containsString("ARG = abc def"));
+			assertThat(out, containsString("ARG = abc;def"));
+		}
+	}
+}

